sql/core/src/main/scala/org/apache/spark/sql/execution/systemcatalog/InformationSchema.scala

```diff
@@ -0,0 +1,312 @@
+    registerView(sparkSession, new ViewsRelationProvider, Seq("views"))
+    registerView(sparkSession, new ColumnsRelationProvider, Seq("columns"))
+    registerView(sparkSession, new SessionVariablesRelationProvider, Seq("session_variables"))
+  }
+
+  /**
+   * Register an INFORMATION_SCHEMA relation provider as a temporary view of Spark Catalog.
+   */
+  private def registerView(
+      sparkSession: SparkSession,
+      relationProvider: SchemaRelationProvider,
+      names: Seq[String]) {
+    val plan =
+      LogicalRelation(relationProvider.createRelation(sparkSession.sqlContext, null, null)).analyze
+    val projectList = plan.output.zip(plan.schema).map {
+      case (attr, col) => Alias(attr, col.name)()
+    }
+    sparkSession.sessionState.executePlan(Project(projectList, plan))
+
+    for (name <- names) {
+      sparkSession.sessionState.catalog.createTempView(s"$INFORMATION_SCHEMA_DATABASE.$name",
+        plan, overrideIfExists = true)
+    }
+  }
+
+  /**
+   * Only EqualTo filters are handled in INFORMATION_SCHEMA data sources.
+   */
+  def unhandledFilters(filters: Array[Filter], columnName: String): Array[Filter] = {
+    import org.apache.spark.sql.sources.EqualTo
+    filters.filter {
+      case EqualTo(attribute, _) if attribute.equalsIgnoreCase(columnName) => false
+      case _ => true
+    }
+  }
+
+  /**
+   * Return `EqualTo` filtered DataFrame.
+   */
+  def getFilteredTables(sparkSession: SparkSession, filters: Seq[Expression], columnName: String)
+      : DataFrame = {
+    import org.apache.spark.sql.catalyst.expressions.EqualTo
+    val database = filters.filter {
+      case EqualTo(AttributeReference(name, _, _, _), Literal(_, StringType))
+        if name.equalsIgnoreCase(columnName) => true
+      case _ => false
+    }.map(_.asInstanceOf[EqualTo].right.asInstanceOf[Literal].eval().toString()).headOption
+
+    val tableList = if (database.nonEmpty) {
+      sparkSession.catalog.listTables(database.get)
+    } else {
+      // Lookup the all table information from all databases.
+      val allTables = new ArrayBuffer[org.apache.spark.sql.catalog.Table]
+      val nonSystemDBs = sparkSession.catalog.listDatabases()
+        .filter(_.name != INFORMATION_SCHEMA_DATABASE).collect()
+      for (db <- nonSystemDBs)
+        allTables ++= sparkSession.catalog.listTables(db.name).collect()
+      CatalogImpl.makeDataset(allTables, sparkSession)
+    }
+
+    tableList.selectExpr("'default'", s"IFNULL(database, '$DEFAULT_DATABASE')", "name",
+      "IF(tableType='VIEW', 'VIEW', IF(isTemporary, 'VIEW', 'TABLE'))",
+      "IFNULL(description, 'VIEW')", "isTemporary"
+    ).toDF("TABLE_CATALOG", "TABLE_SCHEMA", "TABLE_NAME", "TABLE_TYPE", "VIEW_DEFINITION",
+      "isTemporary")
+  }
+}
+
+/**
+ * A base class for all INFORMATION_SCHEMA relation providers.
+ */
+private abstract class InformationSchemaRelationProvider extends SchemaRelationProvider {
```

Comment:
It sounds like you are not passing schema to build the BaseRelation. Why are you extending `SchemaRelationProvider`?